drivers: serial: implement stm32 uart async api

shlomow opened the pull request:
Add initial implementation of the uart async api for stm32 mcus. This uses the dma controller in normal mode for reception. In addition, to detect reception of bytes we enable the idle line detection interrupt.

The testing device is nucleo-l4r5zi.

This PR is based on #14916.
Also, parts of the implementation referred to the implementation from @giansta and @jli157

shlomow committed:
tests: uart_async_api: flush reception in read_abort

After finish transmitting some bytes might be in the
peripheral receive register, since the transmition
is still happening after reception disabled.
This patch ensures that the peripheral has no
bytes in its internal state.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

erwango commented on drivers/serial/Kconfig.stm32:
```diff
@@ -3,10 +3,15 @@
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
+	# the ASYNC implementation requires a DMA controller
+	select SERIAL_SUPPORT_ASYNC if $(dt_compat_enabled,$(DT_COMPAT_STM32_DMA))
```
`st,stm32-dma` could be enabled for other purpose (eg: spi), so we can't assume SERIAL_SUPPORT_ASYNC=y only based on `st,stm32-dma` status.

shlomow replied:
Thanks for the reply! I am not sure I understand. `SERIAL_SUPPORT_ASYNC` only states that the async api is supported and not that it is enabled. For that one needs to enable `UART_ASYNC_API`. This is how it was done in `Kconfig.sam0`

erwango replied:
> SERIAL_SUPPORT_ASYNC only states that the async api is supported and not that it is enabled. For that one needs to enable UART_ASYNC_API. 

Problem remains. Let's say board dts contains a dma node enabled because dma is used on a spi node (or qspi). This doesn't mean dmas prop is set for a uart node and that compilation will be fine.
So I'd prefer to have `SERIAL_SUPPORT_ASYNC` set unconditionally and instead work on a meaningful error is node is misconfigured.
```
select SERIAL_SUPPORT_ASYNC
```

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -621,16 +626,663 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+	/* Configure dma tx config */
+	memset(&data->dma_tx.blk_cfg, 0, sizeof(data->dma_tx.blk_cfg));
+
+#if defined(STM32F1) || defined(STM32F2) ||\
+	defined(STM32F4) || defined(STM32L1)
```
Please use Zephyr flags (`CONFIG_SOC_SERIES_STM32F1X`, ...)

erwango replied:
This request still stands

shlomow replied:
Sorry! I fixed in one place and not in the other.

erwango reviewed (changes requested):
Txs for this contribution!
This looks good in general, but will require some time to review. I made quick initial comments.
Also, what will be missing is an in-tree board updated for this support so that this can be tested (`tests/drivers/uart/uart_async_api`).

FRASTM reviewed (commented):
few rmk else LGTM including @erwango 's comment

shlomow commented:
@erwango @FRASTM Thanks for your replies! I wanted to add an in-tree board in the tests, but since I tested it on nucleo-l4r5zi I need #30784 to be merged.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+			return -ENODEV;
+		}
+	}
+
```
No error if `data->dma_rx.dma_name == NULL` ?

shlomow replied:
No, since it is called in the init function of the async_api which is called in the init of the uart driver. It is valid to not include any dma in uarts that don't want that.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+			LOG_ERR("%s device not found", data->dma_tx.dma_name);
+			return -ENODEV;
+		}
+	}
```
No error if `data->dma_tx.dma_name == NULL` ?

shlomow replied:
No, since it is called in the init function of the async_api which is called in the init of the uart driver. It is valid to not include any dma in uarts that don't want that.

erwango replied:
Indeed, txs

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -749,14 +1393,53 @@ static int uart_stm32_init(const struct device *dev)
+#ifdef CONFIG_UART_ASYNC_API
+	return uart_stm32_async_init(dev);
+#else
+
```
Please remove blank line here

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+
+static void uart_stm32_async_rx_timeout(struct k_work *work)
+{
+	LOG_DBG("rx timeout");
```
Any reason LOG is before local variables definitions ?

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+
+	struct uart_dma_stream *rx_stream = CONTAINER_OF(work,
+			struct uart_dma_stream, timeout_work);
+
```
Please remove blank lines between local variables definitions

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+
+	struct uart_stm32_data *data = CONTAINER_OF(rx_stream,
+			struct uart_stm32_data, dma_rx);
+
```
Please remove blank lines between local variables definitions

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+	if (data->dma_rx.enabled) {
+		LOG_WRN("RX was already enabled");
+		return -EBUSY;
+
```
Please remove blank line

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -626,16 +631,647 @@ static void uart_stm32_irq_callback_set(const struct device *dev,
+		return -EFAULT;
+	}
+
+	USART_TypeDef *UartInstance = UART_STRUCT(dev);
```
Please move this line with other local variables definitions

erwango reviewed (changes requested):
Thanks for this addition. Some few comments bug generally looks good.
Tested ok on nucleo_wb55rg and nucleo_f429zi

Also, I think it would be fair to add `Signed-off-by: Jun Li <jun.r.li@intel.com>` in main commit as @jli157 provided a good base for this.

jli157 commented:
> Thanks for this addition. Some few comments bug generally looks good.
> Tested ok on nucleo_wb55rg and nucleo_f429zi
> 
> Also, I think it would be fair to add `Signed-off-by: Jun Li <jun.r.li@intel.com>` in main commit as @jli157 provided a good base for this.

Thanks or the recognition, @erwango ! Sorry I've been overwhelmed by other things in the past months and didn't finalize the PR I had made. Fortunately, @shlomow is continuing the work and hopefully we can get it done with this great effort! Let me know if I can help on something.

erwango commented on drivers/serial/Kconfig.stm32:
```diff
@@ -7,6 +7,7 @@ config UART_STM32
 	bool "STM32 MCU serial driver"
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
+	select SERIAL_SUPPORT_ASYNC
```
It seems that CI is failing due to misconfiguration.
Given the trouble, I propose you set this part back to your initial proposal:
```
	select SERIAL_SUPPORT_ASYNC if ($(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V1)) || $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2)))
	select DMA if ($(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V1)) || $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2)))
```
And we'll need to find out a better way to handle these dependencies as a latter step.

shlomow committed:
tests: uart_async_api: Support nucleo_l4r5zi board

This commit adds the asynchronous UART API testing support on the
nucleo_l4r5zi board.

The usart3 module is used as the secondary loop-back UART, which is
required to run this test.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow committed:
drivers: serial: implement stm32 uart async api

Add initial implementation of the uart async api
for stm32 mcus. This uses the dma controller
in normal mode for reception. In addition, to detect
reception of bytes we enable the idle line detection
interrupt.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>
Signed-off-by: Jun Li <jun.r.li@intel.com>
Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

fouge commented:
@shlomow @erwango 
Hey
Do you remember which baud rate you've been using to test this?
I am trying to use it at 1Mbps but it seems like the buffer replacement when DMA buffer is full is taking too much time so I am loosing bytes. One operation I want to perform in the callback is to copy the buffer into another one to reconstitute a message. (500+ bytes).
I think we should use a circular buffer instead of the normal one and call `UART_RX_RDY` on half and full buffer so that we never miss any byte as we don't have to manage RX DMA buffers once RX is enabled.

I also tried reloading the DMA config with a new buffer each time the UART goes into idle and I don't miss any byte anymore but the circular buffer is, I think, a better solution.
